### PR TITLE
doc: ipc: Add detailed protocol specs for ICMsg backend

### PR DESCRIPTION
**Pull request by doki-nordic**
The details of how the ICMsg communication is done, are hidden inside the source code. This specification adds details about the ICBMsg protocol.

### CODE REVIEW

**Comment by doki-nordic** [comment]
Direct link to docs:
https://builds.zephyrproject.io/zephyr/pr/78045/docs/services/ipc/ipc_service/backends/ipc_service_icmsg.html#detailed-protocol-specification

File: doc/services/ipc/ipc_service/backends/ipc_service_icmsg.rst [hunk 1]
```diff
@@ -69,18 +69,127 @@ Bonding
...
+For this reason, the specification below focuses on one such pair.
+The other pair is identical.
+
+The ICMsg provides just one endpoint per instance.
+
+Shared Memory Region Organization
+---------------------------------
+
+If data caching is enabled, the shared memory region provided to ICMsg must be
+aligned according to the cache requirement.
```
Comment (divipillai) [inline doc/services/ipc/ipc_service/backends/ipc_service_icmsg.rst:103]:
```suggestion
If data caching is enabled, the shared memory region provided to ICMsg must be aligned according to the cache requirement.
```

File: doc/services/ipc/ipc_service/backends/ipc_service_icmsg.rst [hunk 2]
```diff
@@ -69,18 +69,127 @@ Bonding
...
+If cache is not enabled, the required alignment is 4 bytes.
+
+The shared memory region is entirely used by a single FIFO.
+It contains read and write indexes followed by the data buffer.
+Detailed structure is contained in the following table.
```
Comment (divipillai) [inline doc/services/ipc/ipc_service/backends/ipc_service_icmsg.rst:108]:
```suggestion
Detailed structure is contained in the following table:
```

File: doc/services/ipc/ipc_service/backends/ipc_service_icmsg.rst [hunk 3]
```diff
@@ -69,18 +69,127 @@ Bonding
...
+   * - ``data``
+     - everything to the end of the region
+     - n/a
+     - Circular buffer containing actual bytes to transfer.
+
+This is usual FIFO with a circular buffer, so:
+
+* indexes (``rd_idx`` and ``wr_idx``) are wrapped around when they reach the end of the ``data`` buffer,
+* when ``rd_idx == wr_idx``, the FIFO is empty,
+* it has one byte less capacity than the ``data`` buffer length.
```
Comment (divipillai) [inline doc/services/ipc/ipc_service/backends/ipc_service_icmsg.rst:138]:
```suggestion
This is usual FIFO with a circular buffer, so indexes (``rd_idx`` and ``wr_idx``) are wrapped around under the following conditions:

* When they reach the end of the ``data`` buffer.
* When ``rd_idx == wr_idx``, the FIFO is empty,
* When it has one byte less capacity than the ``data`` buffer length.
```

Please check if the meaning has changed here.

File: doc/services/ipc/ipc_service/backends/ipc_service_icmsg.rst [hunk 4]
```diff
@@ -69,18 +69,127 @@ Bonding
...
+
+Packets
+-------
+
+Packets are sent over the FIFO described in the above section.
+One packet can be wrapped around if it occurs at the end of the FIFO buffer.
+
+The packet structure is following.
```
Comment (divipillai) [inline doc/services/ipc/ipc_service/backends/ipc_service_icmsg.rst:146]:
```suggestion
The following is the packet structure:
```

File: doc/services/ipc/ipc_service/backends/ipc_service_icmsg.rst [hunk 5]
```diff
@@ -69,18 +69,127 @@ Bonding
...
+     - n/a
+     - Packet data.
+   * - ``padding``
+     - 0‑3
+     - n/a
+     - Padding is added to align the total packet size to 4 bytes.
+
+The packet send procedure is following:
+
+#. Check if the packet will fit into the buffer.
```
Comment (divipillai) [inline doc/services/ipc/ipc_service/backends/ipc_service_icmsg.rst:175]:
```suggestion
#. Check if the packet fits into the buffer.
```

File: doc/services/ipc/ipc_service/backends/ipc_service_icmsg.rst [hunk 6]
```diff
@@ -69,18 +69,127 @@ Bonding
...
+   * - ``data``
+     - ``len``
+     - n/a
+     - Packet data.
+   * - ``padding``
+     - 0‑3
+     - n/a
+     - Padding is added to align the total packet size to 4 bytes.
+
+The packet send procedure is following:
```
Comment (divipillai) [inline doc/services/ipc/ipc_service/backends/ipc_service_icmsg.rst:173]:
```suggestion
The packet send procedure is following:
```

File: doc/services/ipc/ipc_service/backends/ipc_service_icmsg.rst [hunk 7]
```diff
@@ -69,18 +69,127 @@ Bonding
...
+#. Check if the packet will fit into the buffer.
+#. Write the packet to ``data`` FIFO buffer starting at ``wr_idx``.
+   Wrap it if needed.
+#. Write a new value of the ``wr_idx``.
+#. Notify the receiver over the MBOX channel.
+
+Initialization
+--------------
+
+The initialization sequence is following:
```
Comment (divipillai) [inline doc/services/ipc/ipc_service/backends/ipc_service_icmsg.rst:184]:
```suggestion
The initialization sequence is the following:
```

File: doc/services/ipc/ipc_service/backends/ipc_service_icmsg.rst [hunk 8]
```diff
@@ -69,18 +69,127 @@ Bonding
...
+#. Set the ``wr_idx`` and ``rd_idx`` to zero.
+#. Push a single packet to FIFO containing magic data: ``45 6d 31 6c 31 4b 30 72 6e 33 6c 69 34``.
+   The MBOX is not used yet.
+#. Initialize the MBOX.
+#. Repeat notification over the MBOX channel using some interval, for example, 1 ms.
+#. Wait for an incoming packet containing the magic data.
+   It will arrive over the other pair (shared memory region and MBOX).
+#. Stop repeating the MBOX notification.
+
+After this, the ICMsg is bounded and it is ready to transfer packets.
```
Comment (divipillai) [inline doc/services/ipc/ipc_service/backends/ipc_service_icmsg.rst:195]:
```suggestion
After this, the ICMsg is bounded, and it is ready to transfer packets.
```

File: doc/services/ipc/ipc_service/backends/ipc_service_icmsg.rst [hunk 9]
```diff
@@ -69,18 +69,127 @@ Bonding
...
+   is read from ``rx-region``. If it is correct, the bonding process is finished
+   and the backend informs the application by calling
+   :c:member:`ipc_service_cb.bound` callback.
 
 Samples
 =======
 
  - :zephyr:code-sample:`ipc-icmsg`
+
+Detailed Protocol Specification
```
[the remark this answers is not in the document]
Comment (doki-nordic) [reply doc/services/ipc/ipc_service/backends/ipc_service_icmsg.rst:88]: I disagree. The "Detailed Protocol Specification" is for users that need to know low level details about the protocol (which is very unusual case). The configuration is for all user. They do not need to know how the shared memory is organized and when MBOXes are used. They just need to assign resources for them using configuration. Shared memory and MBOXes are HW or system specific things (not ICMsg specific), so user should be familiar with them before reading "Detailed Protocol Specification".

File: doc/services/ipc/ipc_service/backends/ipc_service_icmsg.rst (hunk 3 above)
Comment (doki-nordic) [reply doc/services/ipc/ipc_service/backends/ipc_service_icmsg.rst:138]: It is different meaning. I was trying to say that this is usual FIFO, but such statement is not strict enough, so I added those 3 points characterizing it for clarification. Each of them is independent, and you joined them which is wrong. Maybe rewrite the points as follows?
```
This is usual FIFO with a circular buffer:

* The indexes (``rd_idx`` and ``wr_idx``) are wrapped around when they reach the end of the ``data`` buffer.
* The FIFO is empty if ``rd_idx == wr_idx``.
* The FIFO has one byte less capacity than the ``data`` buffer length.
```

File: doc/services/ipc/ipc_service/backends/ipc_service_icmsg.rst (hunk 6 above)
Comment (doki-nordic) [reply doc/services/ipc/ipc_service/backends/ipc_service_icmsg.rst:173]: I see no change there.

File: doc/services/ipc/ipc_service/backends/ipc_service_icmsg.rst (hunk 9 above)
[the remark this answers is not in the document]
Comment (divipillai) [reply doc/services/ipc/ipc_service/backends/ipc_service_icmsg.rst:88]: Agree, thanks for the explanation.

File: doc/services/ipc/ipc_service/backends/ipc_service_icmsg.rst (hunk 3 above)
Comment (divipillai) [reply doc/services/ipc/ipc_service/backends/ipc_service_icmsg.rst]: Agree.

**Review by divipillai — APPROVED** [review]

**Review by maje-emb — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Review by anangl — APPROVED** [review]

**Pull request merged**